packages/common/src/otel.ts

```diff
@@ -0,0 +1,46 @@
+import { isSpanContextValid, Span, SpanContext } from '@opentelemetry/api';
+import { defaultDataConverter } from './converter/data-converter';
+import { Headers } from './interceptors';
+
+export const TRACE_HEADER = 'Otel-Trace-Context';
+
+/**
+ * If found, return a span context deserialized from the provided headers
+ */
+export async function extractSpanContextFromHeaders(headers: Headers): Promise<SpanContext | undefined> {
+  const encodedSpanContext = headers[TRACE_HEADER];
+  if (encodedSpanContext == null) {
+    return undefined;
+  }
+  const decoded: SpanContext = await defaultDataConverter.fromPayload(encodedSpanContext);
+  if (isSpanContextValid(decoded)) {
+    return decoded;
+  }
+  return undefined;
+}
+
+/**
+ * Given a span context, insert it into the provided headers if they exist
+ */
+export async function insertSpanContextIntoHeaders(context: SpanContext, headers?: Headers): Promise<void> {
+  if (headers === undefined) {
+    return;
+  }
+  headers[TRACE_HEADER] = await defaultDataConverter.toPayload(context);
+}
+
+/**
+ * Link a span to an maybe-existing span context
+ */
+export function linkSpans(fromSpan: Span, toContext?: SpanContext): void {
```

Comment:
I think you can set the links on span creation, will that work?

Reply:
No, we don't know the link until after the span is created, which apparently is just generally a hole in the Opentelem spec, as both setting trace ID or setting links are only expected to happen at span creation time, which just misses quite a big section of use cases :|

I might have to change this so that the polling spans don't really get linked and instead the inner spans which are only created after we have received the info about the Run Id are, but that requires more inspecting of the results in places that ideally might be free of that.